src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -963,6 +963,20 @@ private TResult ExecuteProtocol<TResult>(IWireProtocol<TResult> protocol, Cancel
                 }
             }
 
+            private TResult ExecuteProtocol<TResult>(IWireProtocol<TResult> protocol, ICoreSession session, CancellationToken cancellationToken)
+            {
+                try
+                {
+                    return protocol.Execute(_connection, cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    MarkSessionDirtyIfNeeded(session, ex);
+                    _server.HandleChannelException(_connection, ex);
+                    throw;
+                }
+            }
```

Comment:
It is possible to amend overloaded method by specifying `ICoreSession session` as an optional parameter of  existing `ExecuteProtocol<TResult>` method, but considering that `CancellationToken` is usually the last argument in method I would rather stick to overload option.

Reply:
SGTM